Use defined path aliases for prerendering

esteban-url opened the pull request:
fixes #7574 and maybe makes this too #3729

### The problem:

You can configure typescript, webpack, jest and vite to use aliases for your imports, but the prerendering process cannot resolve those paths, and fails when you user them.

### The fix:

1.Read the web `tsconfig.json` file.
1. Extract the paths that were added by the user (aka the aliases). 
1. Add those aliases to the babel config that is used during prerendering.

Please do let me know if there is a better way to do this. I'll be glad to refactor and learn.

Once this gets reviewed and merged, I can write a how-to guide to use path aliases ✨

Tobbe commented on packages/prerender/src/runPrerender.tsx:
```diff
@@ -177,6 +182,17 @@ export const runPrerender = async ({
+          // gets the config from `web/tsconfig.json`
+          // then add the the user defined aliases
+          alias: getPathsFromWebTsconfig(),
+          loglevel: 'silent', // to silence the unnecessary warnings
```
What are the unnecessary warnings?

esteban-url replied:
It will tell me it can't resolve the components, but everything works.

```sh
rw-prerender-aliases|main⚡ ⇒ yarn rw build
✔ Generating Prisma Client...
✔ Generating Prisma Client...
✔ Verifying graphql schema...
✔ Building API...
✔ Cleaning Web...
✔ Building Web...
Starting prerendering...
⠋ Prerendering / -> web/dist/index.html
Could not resolve "src/components/ui/Emoji/Emoji" in file /Users/esteban/files/github/esteban-url/rw-prerender-aliases/web/src/pages/HomePage/HomePage.tsx.
Could not resolve "src/components/ui/Hello/Hello" in file /Users/esteban/files/github/esteban-url/rw-prerender-aliases/web/src/p
✔ Prerendering / -> web/dist/index.html
```

When I leave it on, it's quiet, and everything still works

```sh
rw-prerender-aliases|main⚡ ⇒ yarn rw build
✔ Generating Prisma Client...
✔ Generating Prisma Client...
✔ Verifying graphql schema...
✔ Building API...
✔ Cleaning Web...
✔ Building Web...
Starting prerendering...
⠋ Prerendering / -> web/dist/index.html
Could not resolve "src/components/ui/Emoji/Emoji" in file /Users/esteban/files/github/esteban-url/rw-prerender-aliases/web/src/pages/HomePage/HomePage.tsx.
Could not resolve "src/components/ui/Hello/Hello" in file /Users/esteban/files/github/esteban-url/rw-prerender-aliases/web/src/p
✔ Prerendering / -> web/dist/index.html```

esteban-url replied:
the newest commit got rid of this changes all together. resolving

dac09 commented:
Hey @esteban-url - thanks for the PR!

A few suggestions to get this in:

1. I don't think you need an external `json-c` parser. You can use typescript to parse it like I do here: https://github.com/redwoodjs/redwood/blob/main/packages/internal/src/project.ts#L4-L14
2. You will need to check for both `tsconfig.json` and `jsconfig.json` (For JS projects)
3. You will need to make sure it works for both api and web sides, for consistency. In its current state it only works for the web side I think.
4. If you apply the changes you made to the "base" babel config, it'll automatically configure builds and jest too: 
	- For API Jest: https://github.com/redwoodjs/redwood/blob/main/packages/internal/src/build/babel/api.ts#L104-L109
	- For API build: I think we need to drop the custom redwood-src-alias plugin we have, and instead use the same one we use for Jest here if we want to support this properly
	- For Web: just these settings https://github.com/redwoodjs/redwood/blob/main/packages/internal/src/build/babel/web.ts#L29-L34

Context:
The reason we have the `babel-plugin-module-resolver` as well as our custom src plugin is a bit of techdebt. Without the module-resolver plugin `jest.mock` doesn't work with the aliases!

dac09 reviewed (changes requested):
Requesting a few changes :)

esteban-url commented:
Thanks you so much for the pointers @dac09! I've updated the PR and it handles testing and building automatically by only adding the paths to the `[ts|js]config.json`

dac09 commented:
Thank you so much for this @esteban-url, and apologies for the delay in reviewing this. Your changes look great, and I've verified that module aliases are working correctly on the web side during dev, build and prerender. 

However the api side seems to throw with module aliases, and there's a couple of gotchas that I think we should include in the documentation when you write it - a little difficult to write up so have a recording to explain: https://s.tape.sh/VBlhPCbM?s=1.25

dac09 commented on packages/internal/src/build/babel/common.ts:
```diff
@@ -62,3 +68,65 @@ export const getCommonPlugins = () => {
+
+  const { baseUrl, paths } = config.compilerOptions
+  const pathsObj: Record<string, string> = {}
+  for (const [key, value] of Object.entries(paths)) {
```
Question: do we need to do this? Or can we just remove the predefined stuff in: https://github.com/redwoodjs/redwood/pull/7575/files#diff-9aac922074f3addea47d2d5cddf7eb2fa22adc679d08061f8aab28581bd14430R113

esteban-url replied:
not sure what you mean by removing the predefined stuff there, since we don't remove them, we leave them as they are and exclude them from the transformation, otherwise quite a few things break.

dac09 replied:
👍

jtoar commented:
Hey @esteban-url, do you want to keep working on this one?

dac09 commented:
@esteban-url thank you so much for the work on this, much appreciated! Would you like some help to get this over the line? We can find 30mins or an hour to pair on this real time if that helps!

esteban-url commented:
Hey @jtoar @dac09, thanks for the ping, i def wanna work on this. I'm sorry I dropped the ball, I cant remember why I didn't follow up sooner. 

Anyways, Just updated my fork, and reviewing my code to remember what i was doing 😅. 

Danny, that tape is so insightful! thank you so much, I would love to have a call, but let me have a stab at it first, that way I'll actually know what im looking at(tm)

On the other hand if this is a pressing issue, please don't let me hold it back.

dac09 commented:
All good Esteban, keep going! Just let me know if I can help, my hands are full right now anyway 🙂

esteban-url commented:
ok, so, I think the api side is working now, used the same example you showed in the tape.

the problem was that we were only loading the custom paths when `forJest` was true. removing that one bit to always include the paths seem to have fixed the problem and AFAICT doesn't cause  any side effects. Both the test app and the framework test suites pass

esteban-url commented:
as for the documentation on the mirror types, @dac09 do you think they could be automagically included in the same babel plugin? instead of having to add them manually.

For example, if I add this in any of my `tsconfig.json` files:

```
"@lib/*": [
        "./src/lib/*",
      ],
```

we add the mirror types:

```
"@ui/*": [
        "./src/components/ui/*",
        "../.redwood/types/mirror/web/src/components/ui/*"
      ],
```

would that be worth trying?

esteban-url commented:
this is a rough draft of what the docs could look like, I just don't know what the best place for it is, the `typescript` section maybe? @jtoar any ideas?

----


 # Alias Paths

Alias paths are a mechanism that allows you to define custom shortcuts or aliases for import statements in your code. Instead of using relative or absolute paths to import modules or files, you can use these aliases to make your imports cleaner and more concise.

Redwood comes with a great starting point by aliasing the `src` directory, but you can take this further by configuring your `tsconfig.json` file, your import paths could go from:

```ts
// this really long path
import { CustomModal } from 'src/components/modules/admin/common/ui/CustomModal'

// to this nicer one!
import { CustomModal } from '@adminUI/CustomModal'
```

Add you custom aliases in your `tsconfig.json` file:

```json
{
  "compilerOptions": {
...
    "paths": {
      "src/*": ["./src/*", "../.redwood/types/mirror/api/src/*"],

      "@adminUI/*": [
        "./src/components/modules/admin/common/ui/*",
        "../.redwood/types/mirror/web/src/components/modules/admin/common/ui/*"
      ],

      "types/*": ["./types/*", "../types/*"],
      "@redwoodjs/testing": ["../node_modules/@redwoodjs/testing/api"]
    }
  }
...
}
```

You might have noticed the `"../.redwood/types/mirror/web/src/components/modules/admin/common/ui/*"` path. I'm glad you did! Redwood does a great deal of work to make our lives easier. When you build your project it will create a `.redwood` directory and automagically create index files in each of our directories so we don't have to. These index files allow us to import a module using only the directory where it's located in. 

When you combine those two paths `.src/...` and `./.redwood/...` under an alias you can have shorter and cleaner import paths 

```ts
// Instead of this 🥵
import { CustomModal } from 'src/components/modules/admin/common/ui/CustomModal/CustomModal'

// they could look like this ✨
import { CustomModal } from '@adminUI/CustomModal'
```

## Some benefits of using alias paths are:

1. **Improved code readability**, by abstracting complex directory hierarchies, and having meaningful names for your imports.
1. **Code maintainability**,  aliases allow you to decouple your code from the file structure and more easily move files around, as they are not tied to the longer path.
1. **Reduce boilerplate**, no more `../../src/components/modules/admin/common/ui/` 😮‍💨

dac09 commented:
Looking good @esteban-url! Just need the docs and we should be good to go :), will wait for Dom's suggestions here.

thedavidprice commented:
^^ @jtoar

esteban-url commented:
Thanks for the help @dac09! I've briefly updated the docs, I think they can be better. So I'll wait for @jtoar comments to wrap it all up.

Excited to have the feature built it!

jtoar commented on packages/internal/src/build/babel/web.ts:
```diff
@@ -42,6 +48,8 @@ export const getWebSideBabelPlugins = (
       require('../babelPlugins/babel-plugin-redwood-src-alias').default,
       {
         srcAbsPath: rwjsPaths.web.src,
+        // adds the paths from [ts|js]config.json to the module resolver
+        ...getPathsFromConfig(config.web),
```
I don't think this will do anything; `babel-plugin-redwood-src-alias` is an in-house Babel plugin that re-writes `./src`. It only takes one option, `srcAbsPath`:

https://github.com/redwoodjs/redwood/blob/88f68ef867db29327c0848f9bed5e89435775683/packages/internal/src/build/babelPlugins/babel-plugin-redwood-src-alias.ts#L7-L10 

```suggestion
```

jtoar commented on packages/internal/src/build/babel/web.ts:
```diff
@@ -11,12 +11,16 @@ import {
   { forJest, forVite }: Flags = { forJest: false, forVite: false }
 ) => {
   const rwjsPaths = getPaths()
+  //get the config object from the file
```
Let's improve this comment a bit:

```suggestion
  // Get the TS configs in the api and web sides as an object
```

esteban-url replied:
I'm glad to change this, but the config file can also be a `jsconfig.json` should we still change it to TS configs?

jtoar replied:
You're right but I think it’s ok to refer to it as TS config; they’re both files for configuring TypeScript, so I think TS config gets it across

esteban-url replied:
perfect, thanks for the clarification!

jtoar commented on packages/internal/src/build/babel/common.ts:
```diff
@@ -62,3 +68,65 @@ export const getCommonPlugins = () => {
+ * Finds, reads and parses the [ts|js]config.json file
+ * @returns The config object
+ */
+export const parseConfigFiles = () => {
```
In line with the other comments, let's be more explicit about which config files:

```suggestion
export const parseTypeScriptConfigFiles = () => {
```

esteban-url replied:
this method can also parse `jsconfig.json` files, should we rename it?

jtoar replied:
See https://github.com/redwoodjs/redwood/pull/7575#discussion_r1196940890

jtoar commented on packages/internal/src/build/babel/web.ts:
```diff
@@ -11,12 +11,16 @@ import {
 ) => {
   const rwjsPaths = getPaths()
+  //get the config object from the file
+  const config = parseConfigFiles()
```
`config` isn't too descriptive; can we change it to something like `tsConfigs`? (Plural because this is both right?)

```suggestion
  const tsConfigs = parseConfigFiles()
```

esteban-url replied:
same as the other two comments, this can be regular JS, `jsconfig.json` file

jtoar replied:
See https://github.com/redwoodjs/redwood/pull/7575#discussion_r1196940890

jtoar reviewed (changes requested):
Hey @esteban-url, I tried this locally and it seems to work! I still need to try the api side more thoroughly. But here's a first round of comments. My suggestions are mainly around naming and being more explicit. I'll follow up with another when I get the chance

jtoar commented on packages/internal/src/build/babel/api.ts:
```diff
@@ -105,12 +109,13 @@ export const getApiSideBabelPlugins = (
      *
      */
     ['@babel/plugin-transform-runtime', BABEL_PLUGIN_TRANSFORM_RUNTIME_OPTIONS],
-    // // still needed for jest.mock
-    forJest && [
+    [
       'babel-plugin-module-resolver',
       {
         alias: {
           src: './src',
+          // adds the paths from [ts|js]config.json to the module resolver
+          ...getPathsFromConfig(config.api),
```
I acknowledge the tech debt in these configs, but instead of getting rid of the `forJest` parameter, could we mimic the logic that the web side has where if `forJest` is `true` it conditionally adds in the src alias? 

https://github.com/redwoodjs/redwood/blob/8f672d3fca0b311b68f398a937a721ca93e82a31/packages/internal/src/build/babel/web.ts#L37

esteban-url replied:
I couldn't see the two sides of that condition, one being: add the src alias when `forJest` is true, but we also want it to be there when its false. So I couldn't figure out how to do this.

Adding this would cause the tests to fail, so i decided to leave it, but still very curious to understand what you meant, im pretty sure I missed something.
```
 forJest ? rwjsPaths.api.src : './src', 
```

jtoar reviewed (changes requested):
Hey @esteban-url, one more comment on the `forJest` change for the api side. I know there's a lot of tech debt here and things can be confusing, but I think it's safer to keep the way things work for Jest till we refactor. Let me know if my comment is confusing or if there's something I missed

jtoar commented:
As to where the docs should go; maybe under ["A Few Useful Tips"](https://redwoodjs.com/docs/typescript/introduction#a-few-useful-tips) in the Introduction? I'd also be up for making a new page in the TypeScript section. It doesn't feel like it belongs under any of the existing ones we have

jtoar commented:
Hey @esteban-url, anything I can do to help you finish this one? I'm also happy to make the last few changes to get it in, let me know. Just don't want it to get lost because it's an awesome feature and it's so close

jtoar commented:
Sounds good, thank you for all your work here!
